File: apps/api/internal/models/recruitment_submission.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

// RecruitmentSubmission represents a recruitmentsubmission in the system.
type RecruitmentSubmission struct {
	ID        string         `gorm:"primarykey;size:36" json:"id"`
	RecruitmentID string `gorm:"size:36;index" json:"recruitment_id" binding:"required"`
	Recruitment Recruitment `gorm:"foreignKey:RecruitmentID" json:"recruitment"`
	Name string `gorm:"size:255" json:"name" binding:"required"`
	Nim string `gorm:"size:255" json:"nim"`
	DivisionInterestID string `gorm:"size:36;index" json:"division_interest_id" binding:"required"`
	DivisionInterest Division `gorm:"foreignKey:DivisionInterestID" json:"division_interest"`
	Contact string `gorm:"size:255" json:"contact" binding:"required"`
	CustomAnswers datatypes.JSON `gorm:"type:jsonb" json:"custom_answers"`
	Status string `gorm:"size:255" json:"status" binding:"required"`
	Version   int            `gorm:"not null;default:1" json:"version"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

// BeforeCreate generates a UUID before inserting.
func (m *RecruitmentSubmission) BeforeCreate(tx *gorm.DB) error {
	if m.ID == "" {
		m.ID = uuid.New().String()
	}
	return nil
}

// BeforeUpdate increments Version so offline clients can detect server-side updates.
func (m *RecruitmentSubmission) BeforeUpdate(tx *gorm.DB) error {
	m.Version++
	return nil
}

File: packages/shared/schemas/recruitment-submission.ts
import { z } from "zod";

export const CreateRecruitmentSubmissionSchema = z.object({
  recruitmentId: z.string(),
  recruitment: z.unknown(),
  name: z.string(),
  nim: z.string(),
  divisionInterestId: z.string(),
  divisionInterest: z.unknown(),
  contact: z.string(),
  customAnswers: z.unknown(),
  status: z.string(),
  version: z.number().int(),
});

export const UpdateRecruitmentSubmissionSchema = z.object({
  recruitmentId: z.string().optional(),
  recruitment: z.unknown().optional(),
  name: z.string().optional(),
  nim: z.string().optional(),
  divisionInterestId: z.string().optional(),
  divisionInterest: z.unknown().optional(),
  contact: z.string().optional(),
  customAnswers: z.unknown().optional(),
  status: z.string().optional(),
  version: z.number().int().optional(),
});

export type CreateRecruitmentSubmissionInput = z.infer<typeof CreateRecruitmentSubmissionSchema>;
export type UpdateRecruitmentSubmissionInput = z.infer<typeof UpdateRecruitmentSubmissionSchema>;
